flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -56,7 +56,7 @@
 public class ChannelStateWriterImpl implements ChannelStateWriter {
 
 	private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
-	private static final int DEFAULT_MAX_CHECKPOINTS = 5; // currently, only single in-flight checkpoint is supported
+	private static final int DEFAULT_MAX_CHECKPOINTS = 100; // includes max-concurrent-checkpoints + checkpoints to be aborted (scheduled via mailbox)
```

Comment:
I also thought about that in FLINK-17218, but ultimately couldn't find a reason why any arbitrary large value offers an advantage over the current value. It makes the failing `checkState` more unlikely, but didn't eliminate it.

I also thought about removing the `checkState`, but it helped me to find some issues previously.

TL;DR I'm not convinced that this is a proper fix.

Reply:
The map is already cleared during the regular checkpoint abortion by the task thread but it happens with some delay. I see only two reasons why its size could significantly exceed `max-concurrent-checkpoints`:
1. bug in abort procedure
2. task thread is stuck while netty thread continues to receive new barriers fast

Theoretically, 2nd case is unbounded, but in practice, I didn't observe issues with a value of 5 in `UnalignedITCase` after fixing other issues. I think such a scenario where task thread is not aborting writes would signify some problem and should be investigated.

Besides that, aborting async parts of previous checkpoints upon receiving a new barrier won't work with `max-concurrent-checkpoints` > 1.

Reply:
Just thinking loud.

Previously, the issue was that sources received new RPC triggers while being stuck, which enqueued a ton of mails.
With a checkpointing interval of 100ms, you only need to be stuck for 10s until you enqueue 100 mails and hit the limit.

But I guess, the assumption is that now `notifyCheckpointAborted` is called reliably. And if doesn't, we need to fail probably anyways, since something is broken.

So I guess this solution is as good as it gets. I trade-off between quickly identifying bugs in abort and running potentially into some issues with ultra slow tasks and ultra fast checkpoint barriers.